### ISSUE TITLE
Better support deep nesting with/without Page Bundles

### ISSUE THREAD

**Issue by dustinlacewell** [kaushalmodi/ox-hugo#210]
Hello, I have a list index page `_index.md` and within the list template I am doing `{{ .Content }}` however this includes ALL of the subtrees from this part of the Org file, instead of just the content of the parent.

For example:

```org
* Posts
  :PROPERTIES:
  :EXPORT_HUGO_SECTION: blog
  :HUGO_MENU: :menu main
  :EXPORT_FILE_NAME: _index
  :END:

This is a test ?!

** FIRST POST
:PROPERTIES:
:EXPORT_FILE_NAME: first-post
:END:
THIS is the first post!
*** First things first!
blah!
** SECOND POST
:PROPERTIES:
:EXPORT_FILE_NAME: second-post
:END:
HMMM
```

I would like `This is a test ?!` to appear on the index page followed by the itemized posts, but instead ALL of the textual content of the subtrees ends up in `{{ .Content }}` which is probably never useful? Would it make more sense to only pass the specific content of the tree node? I think so.

The template looks like:

```html
<section class="container list">
  <h1 class="title">{{ .Title }}</h1>
  {{ .Content }}
  <ul>
    {{ range .Paginator.Pages }}
    <li>
      <span>{{ .Date.Format "January 2, 2006" }}</span><a href="{{ .URL }}">{{ .Title }}</a>
    </li>
    {{ end }}
  </ul>
  {{ partial "pagination.html" . }}
</section>
```

And it looks like:

![image](https://user-images.githubusercontent.com/53952/46009608-b4b8e680-c085-11e8-8bfb-e9cf2c3430cb.png)

**Comment by kaushalmodi** [comment]
The simple rule to follow is that **a post subtree** (be it a regular single post or list index post) cannot *contain* other posts..

So reorganize the structure as follows and it will work:


```org
* Posts
  :PROPERTIES:
  :EXPORT_HUGO_SECTION: blog
  :HUGO_MENU: :menu main
  :END:
** Title that you want for the Posts list page
  :PROPERTIES:
  :EXPORT_FILE_NAME: _index
  :END:

This is a test ?!

** FIRST POST
:PROPERTIES:
:EXPORT_FILE_NAME: first-post
:END:
THIS is the first post!
*** First things first!
blah!
** SECOND POST
:PROPERTIES:
:EXPORT_FILE_NAME: second-post
:END:
HMMM
```

**Comment by dustinlacewell** [comment]
Wait, so does this mean that we can only ever have one level of content? I couldn't have an "Articles" section, each of which has a chapter, and each chapter has a section, with the ability to have chapter listings and on a given chapter, section listings? And so on?

**Comment by kaushalmodi** [comment]
> Org's most prominent properties as a hierarchical markup isn't being fully utilized in ox-hugo

I don't see it that way. Org does not dictate how the exporter should interpret the subtree hierarchy. In fact, none of the existing exporters even use the tag and property hierarchy as I do in `ox-hugo`. I get your point that if you were to decide how to interpret the hierarchies, you would have used it to translate to directory structure rather than propagate meta-data.

> I worry this will not scale beyond some simple site structures. 

Have you looked at `EXPORT_HUGO_BUNDLE`? Even if you did not know and use that, you could have set the `EXPORT_HUGO_SECTION` to `articles/first-article` in a parent subtree and let that propagate through, instead of repeating the `first-article/` prefix in each `EXPORT_FILE_NAME`.

But you can ignore what I just said :)

Please see the below example that uses `EXPORT_HUGO_BUNDLE` as it should be.

> My intuition is that the Org-mode hierarchy should determine more directly the structure of the site itself.

As I said in Gitter, it is not that simple. 

You need to prepare a detailed RFC that shows (but let's first discuss using the fixed version of your example that I present at the end):

- How you would plan to have the Org subtree hierarchy translate to a directory structure **while not breaking the current tag/property inheritance**.
- How would that directory structure interpretation work with user's `EXPORT_HUGO_SECTION`?
- How would that implementation handle page bundle vs non page bundles?

> I feel like you should be able to structure your site using org-mode's hierarchy, and then use tag and property inheritance as normal, handling special cases with Org's flexibility.

The feature request is too complex. Let's first discuss the below example. I don't see how following that example makes it difficult to make multi-level nesting possible. But may be I am not understanding your **real** use-case.

---

## Nesting example

**Note**: I have simply changed the section name to **ox-hugo-issue-210** from **articles** because I already had a section named that in the sandbox site where I deployed your example.

- [Org file][0] - You can see that in this refactored version, I don't need to repeat "first-article" and "second-article" at multiple places. They happen only once each.
- [Exported Markdown files][4]
- [**ox-hugo-issue-210** landing page][1]
- [first article][2]
- [second article][3]

I you can share your real use case of nesting, I can help you structure it in Org the best way ox-hugo can handle.

[0]: https://gitlab.com/kaushalmodi/hugo-sandbox/raw/master/content-org/ox-hugo-issue-210.org
[1]: https://hugo-sandbox.netlify.com/ox-hugo-issue-210/
[2]: https://hugo-sandbox.netlify.com/ox-hugo-issue-210/first-article/
[3]: https://hugo-sandbox.netlify.com/ox-hugo-issue-210/second-article/
[4]: https://gitlab.com/kaushalmodi/hugo-sandbox/tree/master/content/ox-hugo-issue-210

**Comment by kaushalmodi** [comment]
Here's another example with one more level of nesting.

- [Org file][0]
- [Exported Markdown files][1]
- [**ox-hugo-issue-210b** landing page][2]

[0]: https://gitlab.com/kaushalmodi/hugo-sandbox/raw/master/content-org/ox-hugo-issue-210b.org
[1]: https://gitlab.com/kaushalmodi/hugo-sandbox/tree/master/content/ox-hugo-issue-210b
[2]: https://hugo-sandbox.netlify.com/ox-hugo-issue-210b/

**Comment by dustinlacewell** [comment]
> Even if you did not know and use that, you could have set the EXPORT_HUGO_SECTION to articles/first-article in a parent subtree and let that propagate through, instead of repeating the first-article/ prefix in each EXPORT_FILE_NAME.

How is this possible though? Wont using a third-level sub-tree then cause the `.Content` for the org tree where `EXPORT_FILE_NAME = first-article` will include all of the textual content of the sub-trees. The problem seems to arise when you want to have index pages at each level with custom content. 

The solution to one level of nesting was to put the index and the child pages on the same org-level, so that the `.Content` of the index page wouldn't include the children since they are not sub-trees but siblings in the org document. But then what if those children pages have children of their own? We still can't use additional org sub-tree levels because at any point, nesting will cause the `.Content` of a parent page to contain the text content of all of the sub-trees.

Can you show an example of how you see this working? 

I'm probably just super confused and will continue to read through all information you've provided.

**Comment by kaushalmodi** [comment]
> How is this possible though? 

The example in my last comment https://github.com/kaushalmodi/ox-hugo/issues/210#issuecomment-424565613 shows nested branch bundles. Go though it and let me know if that answers your concerns.

*You cannot nest __leaf__ bundles, but you can nest __branch__ bundles.*

> The solution to one level of nesting was to put the index and the child pages on the same org-level

Not necessarily the same level.. simply remember that you cannot have a subtree with EXPORT_FILE_NAME **inside** another subtree with EXPORT_FILE_NAME too.

>  But then what if those children pages have children of their own? 

That last Org file example covers this case (again, only branch bundles and sections can have children).

>  We still can't use additional org sub-tree levels because at any point, nesting will cause the .Content of a parent page to contain the text content of all of the sub-trees.

The beauty of the current implementation is that you can. Just ensure to not nest anything inside a subtree with EXPORT_FILE_NAME.

> I'm probably just super confused and will continue to read through all information you've provided.

Probably :) But please do get back with comments on the Org examples I gave earlier.

**Comment by dustinlacewell** [comment]
I have read through your org example and it took me about 10 minutes to fully understand the structure of everything.  There is a ton of content and property boilerplate to make a simple three level tree structure and when you just look at the document as a whole, it does not communicate, in my opinion, the structure of the website at all. You have to essentially reverse engineer the content of the document to understand what you're going to get. All so categories and tags can inherit in a specialized way? Any Org user could imagine more natural affairs:

```
* Articles

This is an index page for all the articles

** Article One

This is an index page for Article One chapters.

*** Chapter One

This is an index page for Chapter One sections.

**** Section 1A

This is the content of Section 1A

**** Section 1B

This is the content of Section 1B

*** Chapter Two

This is an index page for Chapter One sections.

**** Section 2A

This is the content of Section 2A

**** Section 2B

This is the content of Section 2B

** Article Two

This is an index page for Article One chapters.

*** Chapter One

This is an index page for Chapter One sections.

....
```

I'm unsure how this structure would make working with tags and property inheritance much worse. But it would make writing and reading (maintaining) such a document much easier when I compare this example to the content of your org example. I think for now I'm going to keep looking for different solutions, but I may end up back here as Org based blogging is my goal. Sincerely, thanks for taking the time to explicate all that information to me.

**Comment by kaushalmodi** [comment]
@dustinlacewell With the latest commit, this is possible (below if your example framework converted to a real exportable Org snippet):

```org
** Hugo Bundle concatenation :fragment:concatenation:deep_nesting:page_bundles:
:PROPERTIES:
:EXPORT_HUGO_SECTION: bundle-concat
:END:
*** Articles
:PROPERTIES:
:EXPORT_HUGO_SECTION*: articles
:END:
**** Articles Index page
:PROPERTIES:
:EXPORT_FILE_NAME: _index
:END:
This is an index page (/branch/ bundle) for all the articles.

[[https://github.com/kaushalmodi/ox-hugo/issues/210#issuecomment-424891371][Ref]]
**** Article One
:PROPERTIES:
:EXPORT_HUGO_BUNDLE: article-one
:END:
***** Article One Index page
:PROPERTIES:
:EXPORT_FILE_NAME: _index
:END:
This is an index page (/branch/ bundle) for Articles/Article One.
***** Chapter One
:PROPERTIES:
:EXPORT_HUGO_BUNDLE: chapter-one
:END:
****** Article One Chapter One Index page
:PROPERTIES:
:EXPORT_FILE_NAME: _index
:END:
This is an index page (/branch/ bundle) for Articles/Article
One/Chapter One.
****** Section 1A
:PROPERTIES:
:EXPORT_FILE_NAME: section-1a
:END:
This is the content of Articles/Article One/Chapter One/Section 1A.
****** Section 1B
:PROPERTIES:
:EXPORT_FILE_NAME: section-1b
:END:
This is the content of Articles/Article One/Chapter One/Section 1B.
```

See https://ox-hugo.scripter.co/doc/hugo-section/ for more info. I have enabled similar property concatenation for `EXPORT_HUGO_BUNDLE` too (HUGO_BUNDLE is yet to be documented: https://github.com/kaushalmodi/ox-hugo/issues/217).

---

The latest few commits resolve most of your queries except two:

1. You have:

    ```
    *** Chapter One
    This is an index page for Chapter One sections.
    **** Section 1A
    This is the content of Section 1A
    ```

    That's not possible.. Org cannot know if Section 1A is a sub-heading in Chapter One or beginning of a new post subtree. So I have this rule that I mentioned earlier:

    > The simple rule to follow is that a post subtree (be it a regular single post or list index post) cannot contain other posts..

2. I don't like to auto-derive section path or exported file names from the subtree heading. And so you need `EXPORT_HUGO_SECTION`/`EXPORT_HUGO_SECTION*`/`EXPORT_FILE_NAME`. If you wish, you can make a Yasnippet or something to auto-derive those properties using `org-hugo-slug` function I define in `ox-hugo`. See [this][1] as an example of how I auto-set `EXPORT_HUGO_BUNDLE` using `org-hugo-slug` in an Org Capture template.

[1]: https://ox-hugo.scripter.co/doc/org-capture-setup/#org-capture-template-page-bundle

**Comment by kaushalmodi** [comment]
Closing out this issue. I believe it's resolved, but feel free to open a new issue if you'd like to discuss further based on my last comment above.
